### DISCUSSION TITLE
I want to understand how to upload files to GitHub…

### DISCUSSION THREAD

**Discussion by rchrdnsh** [sveltia/sveltia-cms#50] [Q&A] [answered]
So this cms uses file uploading to GitHub...I think?... rather than using node js or an external database, etc...which is super interesting to me and what initially attracted me to it...I want to understand how that works and how I could implement something like that myself, but I don't even know where to start. I wanted to start by simply grabbing data from a form, including images, and upload that to a GitHub repo via sveltekit, using any host, like Vercel, netlify, etc...rather than using adapter-node, for instance...

So ultimately my question is how can I do something like that from a svelte or sveltekit app that is not using adapter-node, meaning I would need to communicate with GitHub directly...I think...

For example, how does Sveltia CMS do this?

Anyway, thank you :-)

**Comment by kyoshino** [comment, marked as answer]
Yes, Sveltia CMS uses the GitHub GraphQL API to upload files. [This example](https://github.blog/changelog/2021-09-13-a-simpler-api-for-authoring-commits/) just shows a simple text file, but any file, including images, can be uploaded if encoded in base64. Sveltia’s implementation can be found in [github.js](https://github.com/sveltia/sveltia-cms/blob/main/src/lib/services/backends/github.js#L267-L320) 🙂 I think Netlify/Decap CMS still uses the REST API to upload files, which is [unnecessarily complicated](https://dev.to/bro3886/create-a-folder-and-push-multiple-files-under-a-single-commit-through-github-api-23kc). The GraphQL API is way easier!

**Comment by rchrdnsh** [comment]
hmmmm...could i basically use the github.js file as is in another sveltekit based cms? With the imports, of course. I'm really starting to get interested in the idea of a headless CMS from the perspective of data storage and retrieval...what i mean is the cms is not designed or styled by default, but has built data handling for different data storage options, like github, or supabase(postgres), mongo, etc... maybe like...sveltekit adapters...so the cms infra would be providing different adapters for different data storage options...maybe one could even mix and match them in a form...

So a devloper can design and style the cms in any way they see fit for the project, allowing one to create bespoke custom cms experiences for each client, rather than an off the shelf fully built out cms with limited customization options, but the underlying infra, 
( which is the toughest part to me! XD ) is already there...
 
 I'm really into this idea at the moment... dunno how practical it is, though...

**Reply by kyoshino** [reply]
It all depends on your project! There are actually [many headless CMS options](https://jamstack.org/headless-cms/) in the market. If the project is going to scale, an API-driven (database-backed) solution may be better than a Git-based one. I’ve created Sveltia CMS simply because my clients wanted a drop-in replacement for Netlify CMS 😆

**Reply by kyoshino** [reply]
Importing (JavaScript files inside) Sveltia CMS as an npm package is not yet supported, but I already have a WIP branch for #42. Should be done in a few weeks.

**Reply by rchrdnsh** [reply]
hmmm...i don't mean headless CMS in the way most mean it (decoupled from the website) ...I mean headless CMS in the sense that the CMS has no UI...only the underlying infrastructure for file handling...kind of like Melt UI, the svelte UI library:

https://www.melt-ui.com/

...which is using this concept of builders, so you make your own ui and then hook it up to the functionality the library offers.

Dunno if that makes much sense or not, but from a front end person, such as myself, that part is easy and where I want the control :-)

**Reply by kyoshino** [reply]
Something like this? https://webflow.com/cms-website I have never tried Webflow but I have heard of it for a while.

**Reply by rchrdnsh** [reply]
well, kinda, but I wouldn't need a commercial GIU to make the GUI :-)

I think the idea in my head is more akin to...maybe like utility functions that hook into db's or the file system or github, etc... so one can build a custom cms design from scratch then grab these utilities to talk to these different storage solutions...but i suppose in spirit the webflow thing is similar :-)
